cgen: fix unnecessary default initialization of locals

zerbina opened the pull request:
## Summary

A local of certain types was unnecessarily default-initialized even if
it was assigned immediately (that is, `var x = y`). This doesn't change
observable behaviour, but it does speed up the generated code a bit.

## Details

* only emit default-initialization for definitions when there's no
  initial-value expression
* remove the `isAssignedImmediately` procedure. It's only used a
  single time, and its general name might cause (mis-)use for purposes
  it's not intended for  
* remove the outdated comment in `cgen.initLocalVar`

zerbina committed:
cgen: fix unnecessary default initialization of locals

Summary
=======

A local of certain types was unnecessarily default-initialized even if
it was assigned immediately (that is, `var x = y`). This doesn't change
observable behaviour, but it does speed up the generated code a bit.

Details
=======

* for the definition statement of a local, only consider the local as
  not assigned immediately if it has no initializer expression
* remove the outdated comment in `cgen.initLocalVar`

saem commented on compiler/backend/ccgstmts.nim:
```diff
@@ -16,13 +16,11 @@ const
+  # note: the destination can be considered assigned immediately even if the
+  # source is an exception-raising RVO call. This is because the call happens
+  # *before* the definition, meaning that nothing on the exceptional control-
+  # flow path must observe (read or write) the destination
```
Here "exceptional control flow path" is exclusive of both prior to an exception or in a finally? Basically, one of those latter two needs to write to the destination in order for it to be initialized?

saem reviewed (approved):
A minor question, otherwise it looks good.

saem commented:
/merge

Thought about it a bit more and I don't think the answer to the questions would change anything about this PR, can be discussed in a non-blocking fashion.

zerbina commented:
@saem: Sorry, I was away for a moment. I've also thought about it some more and I do think the comment is a bit unclear, which is partly due to where it's located (it's in a procedure named `isAssignedImmediately` and talks about "definition").

I'll move it somewhere else and will also try to make it a bit clearer.

zerbina committed:
ccgstmts: remove `isAssignedImmediately`

Also add a comment to `genSingleVar` to document why `cnkCheckedCall`
can be ignored.

saem commented:
> @saem: Sorry, I was away for a moment. I've also thought about it some more and I do think the comment is a bit unclear, which is partly due to where it's located (it's in a procedure named `isAssignedImmediately` and talks about "definition").
> 
> I'll move it somewhere else and will also try to make it a bit clearer.

Thanks, I re-read it and figured the code wouldn't really change so started the merge. Totally cool if you want to cancel it and restart it after the change.

zerbina commented:
/merge
